org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,244 @@
+            reversePackDescriptionIndex.put(newPackKey, description);
+            return new DfsPackFile(this, description, newPackKey);
+        });
+    }
+
+    int getBlockSize() {
+        return blockSize;
+    }
+
+    // do something when the block is invalid
+    DfsBlock getOrLoad(DfsPackFile pack, long position, DfsReader dfsReader) throws IOException {
+        final long requestedPosition = position;
+        position = pack.alignToBlock(position);
+        DfsPackKey key = pack.key;
+
+        Ref<DfsBlock> loadedBlockRef = dfsBlockAndIndicesCache.get(new DfsPackKeyWithPosition(key, position), keyWithPosition -> {
+            try {
+                DfsBlock loadedBlock = pack.readOneBlock(keyWithPosition.getPosition(), dfsReader);
+                key.cachedSize.getAndAdd(loadedBlock.size());
+                return new Ref(keyWithPosition.getDfsPackKey(), keyWithPosition.getPosition(),
+                        loadedBlock.size(), loadedBlock);
+            } catch (IOException e) {
+                return null;
+            }
+        });
+
+        if (loadedBlockRef != null) {
+            DfsBlock loadedBlock = loadedBlockRef.get();
+            if (loadedBlock != null && loadedBlock.contains(key, position)) {
+                return loadedBlock;
+            }
+        }
+
+        // the current block is not valid, remove it and attempt `getOrLoad` again
+        dfsBlockAndIndicesCache.invalidate(new DfsPackKeyWithPosition(key, position));
+        return getOrLoad(pack, requestedPosition, dfsReader);
+    }
+
+    void put(DfsBlock block) {
+        put(block.pack, block.start, block.size(), block);
+    }
+
+    public <T> Ref<T> put(DfsPackKey key, long position, int size, T value) {
+        return dfsBlockAndIndicesCache.get(new DfsPackKeyWithPosition(key, position), keyWithPosition -> {
+            // if it's not an index, increase the cached size
+            if (keyWithPosition.position >= 0) {
+                keyWithPosition.getDfsPackKey().cachedSize.getAndAdd(size);
+            }
+            return new Ref(keyWithPosition.getDfsPackKey(), keyWithPosition.getPosition(), size, value);
+        });
+    }
+
+    boolean contains(DfsPackKey key, long position) {
+        return get(key, position) != null;
+    }
+
+    @SuppressWarnings("unchecked")
+    <T> T get(DfsPackKey key, long position) {
+        Ref<T> blockCache = dfsBlockAndIndicesCache.getIfPresent(new DfsPackKeyWithPosition(key, position));
+        return blockCache != null ? blockCache.get() : null;
+    }
+
+    void remove(DfsPackFile pack) {
+        if (pack != null) {
+            DfsPackKey key = pack.key;
+            cleanUpIndicesIfExists(key);
+            key.cachedSize.set(0);
+        }
+        // TODO: release all the blocks cached for this pack file too
+        // right now those refs are not accessible anymore and will be evicted by caffeine cache eventually
```

Comment:
don't see this causing a big problem.... but nice to have it as improvement soon